### PR TITLE
Add support for endTime

### LINKED ISSUE #179: Make endTime work
> Should probably make a point that, while `endTime` is documented, it doesn't affect anything; i.e., there's no actual code that does anything with it if received.

_Originally posted by @Devoxin in https://github.com/Frederikam/Lavalink/issues/170#issuecomment-473971334_

### PR DESCRIPTION
**Pull request by sh30801**
This PR fixes #179, and makes end time functional.

The only thing I am not sure about is handling the seek portion. Right now it will end the track if the seek position is greater than the endtime position, but perhaps it should just cancel the marker and keep the track playing.

### CODE REVIEW

File: LavalinkServer/src/main/java/lavalink/server/io/WebSocketHandlers.kt [hunk 1]
```diff
@@ -47,6 +48,16 @@ class WebSocketHandlers(private val contextMap: Map<String, SocketContext>) {
             player.setVolume(json.getInt("volume"))
         }
 
+        if (json.has("stopTime")) {
```
Comment (devoxin) [inline LavalinkServer/src/main/java/lavalink/server/io/WebSocketHandlers.kt]: As per the [IMPLEMENTATION.md](https://github.com/Frederikam/Lavalink/blob/master/IMPLEMENTATION.md#outgoing-messages), this should be `endTime`.

File: LavalinkServer/src/main/java/lavalink/server/io/WebSocketHandlers.kt [hunk 2]
```diff
@@ -47,6 +48,16 @@ class WebSocketHandlers(private val contextMap: Map<String, SocketContext>) {
...
+
+            val stopTime = json.getLong("stopTime")
```
Comment (devoxin) [inline LavalinkServer/src/main/java/lavalink/server/io/WebSocketHandlers.kt]: See above comment.

**Review by devoxin — CHANGES_REQUESTED** [review]

**Comment by devoxin** [comment]
> The only thing I am not sure about is handling the seek portion. Right now it will end the track if the seek position is greater than the endtime position, but perhaps it should just cancel the marker and keep the track playing

Lavaplayer automatically handles this. When a seek is requested that exceeds the marker position, the marker handler will be invoked with a state denoting so. I believe this should already be handled [by the `BYPASSED` check](https://github.com/Frederikam/Lavalink/blob/master/LavalinkServer/src/main/java/lavalink/server/player/TrackEndMarkerHandler.java#L37), where `BYPASSED` is the state provided when the marker position is exceeded, as [denoted here](https://github.com/sedmelluq/lavaplayer/blob/master/main/src/main/java/com/sedmelluq/discord/lavaplayer/track/TrackMarkerHandler.java#L29).

My only concern is the MarkerHandler line linked above (L37).
`if (state.equals(MarkerState.REACHED) | state.equals(MarkerState.BYPASSED))` should perhaps be `if (state.equals(MarkerState.REACHED) || state.equals(MarkerState.BYPASSED))` instead.

**Review by freyacodes — COMMENTED** [review]
The code looks fine, but has this been tested?

**Comment by sh30801** [comment]
> 
> 
> The code looks fine, but has this been tested?

Yes it has. It works fine both with and without and endtime. Passing in an endtime ends it as expected.

**Review by freyacodes — APPROVED** [review]
:ok_hand:

**Pull request merged**